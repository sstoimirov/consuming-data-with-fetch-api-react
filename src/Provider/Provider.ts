import { observable, action } from "mobx"
import { UserType, TodoType } from "../types";
export class Provider {
    @observable users: UserType[];
    @observable todos: TodoType[];
    constructor() {
        this.users = [];
        this.todos = [];
        this.fetchTodos();
        this.fetchUsers();
    }

    @action.bound setUsers(data: UserType[]) {
        this.users = data;
    }

    @action.bound setTodos(data: TodoType[]) {
        this.todos = data;
    }

    fetchUsers() {
        return fetch('http://jsonplaceholder.typicode.com/users')
            .then(res => res.json())
            .then(data => {
                this.setUsers(data);
            })
            .catch(console.log)
    }

    fetchTodos() {
        return fetch('http://jsonplaceholder.typicode.com/todos')
            .then(res => res.json())
            .then(data => {
                this.setTodos(data);
            })
            .catch(console.log)
    }
}