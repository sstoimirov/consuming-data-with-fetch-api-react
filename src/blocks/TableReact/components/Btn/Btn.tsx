import * as React from 'react';

export class Btn extends React.PureComponent<{ text: string, onClick: (e: React.MouseEvent<HTMLButtonElement>) => void; }, {}>{
    render() {
        return (
            <button type="button" onClick={this.props.onClick}>{this.props.text}</button>
        )
    }
}