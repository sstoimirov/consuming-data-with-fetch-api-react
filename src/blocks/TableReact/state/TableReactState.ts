import { Provider } from "../../../Provider";
import { Dict, dict } from "../dict";
import { sortType, UserType, TodoType } from "../../../types";
import { action, computed, observable, Lambda, observe } from "mobx";
import { sortUsers } from "../../../helpers/sorting"

export class TableReactState {
    private provider: Provider;
    dict: Dict
    @observable sortOrder: sortType;
    @observable users: UserType[]
    @observable todos: TodoType[];
    @observable inputValue: string;
    @observable optionValue: string;

    private disposeUsers: Lambda;
    private disposeTodos: Lambda;
    constructor(provider: Provider) {
        this.provider = provider;
        this.dict = dict;
        this.sortOrder = "ASC";
        this.users = provider.users;
        this.todos = provider.todos;

        this.disposeUsers = observe(this.provider, "users", u => {
            this.setUsers(u.newValue);
        });

        this.disposeTodos = observe(this.provider, "todos", t => {
            this.setTodos(t.newValue);
        })
        this.inputValue = "";
        this.optionValue = "name";
    }

    @action.bound setUsers(u: UserType[]) {
        this.users = u
    }
    @action.bound setTodos(t: TodoType[]) {
        this.todos = t
    }
    @action.bound setOrder(s: sortType) {
        return this.sortOrder = s
    }
    @action.bound updateOptionValue(el: React.FormEvent<HTMLSelectElement>) {
        this.optionValue = el.currentTarget.value
    }
    @action.bound updateInputValue(e: React.FormEvent<HTMLInputElement>) {
        this.inputValue = e.currentTarget.value.toUpperCase();
    }

    // @computed get todosData() {
    //     return this.todos.sort(sortTodos[this.sortOrder])
    // }
    @computed get usersData() {
        return this.users.sort(sortUsers[this.sortOrder])
    }
    deactivate() {
        this.disposeUsers();
        this.disposeTodos();
    }
}