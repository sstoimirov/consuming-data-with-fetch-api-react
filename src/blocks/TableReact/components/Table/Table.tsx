import * as React from 'react';
import { observer } from "mobx-react";
import { TableReactState } from '../../state';
import { UserType, sortType } from '../../../../types';
import * as CMP from "../../components";
@observer
export class Table extends React.PureComponent<{ dataProvider: TableReactState }, TableReactState>{
    constructor(props) {
        super(props)
        this.state = this.props.dataProvider
    }
    getUsers() {
        const allowedKeys = ["name", "username", "email"];
        return (
            this.state.usersData.map((user: UserType) => (
                <div key={`${user.id}`} className="user">
                    {Object.keys(user)
                        .filter(key => allowedKeys.includes(key))
                        .filter(x => user[this.state.optionValue].toUpperCase().indexOf(this.state.inputValue) > -1)
                        .map(key => {
                            return <div key={key} className="user-info">{`${user[key]}`}</div>
                        })
                    }
                </div>
            ))
        )
    }

    // getTodos() {
    //     return (
    //         this.state.todosData.map((todo: TodoType) => (
    //             <div key={`${todo.id}`} className="todo">
    //                 <div className="todo-info">Id: {todo.id}</div>
    //                 <div className="todo-info">Title: {todo.title}</div>
    //             </div>
    //         ))
    //     )
    // }

    getUserKeys() {
        let userKeys: UserType[] = [], keys, j
        const allowedKeys = ["name", "username", "email"];

        for (let i = 0; i < this.state.usersData.length; i++) {
            keys = Object.keys(this.state.usersData[i]).filter(x => allowedKeys.includes(x));
            for (j = 0; j < keys.length; j++) {
                if (userKeys.indexOf(keys[j]) === -1) {
                    userKeys.push(keys[j])
                }
            }
        }
        return userKeys.map(x => <div key={`key__${x}`} className="user-key-value">{x}</div>)
    }

    render() {
        const selectMenuProps = {
            onChange: this.state.updateOptionValue,
            name: this.state.dict.Name,
            userName: this.state.dict.UserName,
            email: this.state.dict.Email
        }
        return (
            <div className="table-wrapper">
                <div className="table-buttons">
                    <CMP.Search text={`Search items by ${this.state.optionValue}`} onChange={this.state.updateInputValue} />
                    <CMP.SelectDropdown {...selectMenuProps} />

                    <CMP.Btn onClick={e => {
                        let sort: sortType = this.state.sortOrder === "ASC" ? "DESC" : "ASC"
                        this.state.setOrder(sort)
                    }} text={`Sort by ${this.state.sortOrder}`} />
                </div>

                <div className="table-info">
                    <div className="users-wrapper">
                        <h2>{this.state.dict.Users}</h2>

                        <div className="user-keys-wrapper">
                            {this.getUserKeys()}
                        </div>

                        {this.getUsers()}
                    </div>

                    {/* <div className="todos-wrapper">
                        <h2>{this.state.dict.Todos}</h2>
                        {this.getTodos()}
                    </div> */}
                </div>
            </div>

        )
    }
}