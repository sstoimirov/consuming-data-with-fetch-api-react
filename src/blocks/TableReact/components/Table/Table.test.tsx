import * as React from "react";
import * as fetch from "jest-fetch-mock";

import { shallow, configure } from 'enzyme';
import { Table } from "./Table";
import Adapter from 'enzyme-adapter-react-16'
import { Provider } from "../../../../Provider";
import { TableReactState } from "../../state";
import { Btn } from "../Btn";
configure({ adapter: new Adapter() })
describe("Table", () => {
  const fetchMock = fetch as fetch.FetchMock;
  let state: TableReactState, provider, onResponse, onError, comp, btn

  beforeEach(() => {
    fetchMock.resetMocks();
    state = new TableReactState(new Provider());
    provider = new Provider();
    onResponse = jest.fn();
    onError = jest.fn();
    comp = shallow(<Table dataProvider={state} />)
    btn = comp.find(Btn).at(0);
  });

  it('Table renders data correclty with provided props for Users and sort DESC', async (done) => {
    fetchMock.mockResponseOnce(JSON.stringify([{ id: 1 }]));
    return await provider.fetchUsers()
      .then(onResponse)
      .catch(onError)
      .finally(() => {
        state.sortOrder = "DESC";
        btn.simulate("click");

        expect(onResponse).toHaveBeenCalled();
        expect(onError).not.toHaveBeenCalled();
        expect(state.sortOrder).toEqual("ASC")
        expect(comp.debug()).toMatchSnapshot()
        done()
      });
  });

  it('Table renders data correclty with provided props for Users and sort ASC', async (done) => {
    fetchMock.mockResponseOnce(JSON.stringify([{ id: 1 }]));
    return await provider.fetchUsers()
      .then(onResponse)
      .catch(onError)
      .finally(() => {
        state.sortOrder = "ASC";
        btn.simulate("click");

        expect(onResponse).toHaveBeenCalled();
        expect(onError).not.toHaveBeenCalled();
        expect(state.sortOrder).toEqual("DESC");
        expect(comp.debug()).toMatchSnapshot();
        done();
      });
  });

  it('Table renders data correclty with provided props for Todos', async (done) => {
    fetchMock.mockResponseOnce(JSON.stringify([{ id: 1 }]));
    const provider = new Provider();
    const onResponse = jest.fn();
    const onError = jest.fn();
    const comp = shallow(<Table dataProvider={state} />)
    return await provider.fetchTodos()
      .then(onResponse)
      .catch(onError)
      .finally(() => {
        expect(onResponse).toHaveBeenCalled();
        expect(onError).not.toHaveBeenCalled();
        expect(comp.debug()).toMatchSnapshot();
        done();
      });
  });
})