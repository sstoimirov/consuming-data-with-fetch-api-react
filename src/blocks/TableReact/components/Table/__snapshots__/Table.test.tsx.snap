// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Table renders data correclty with provided props for Todos 1`] = `
"<div className=\\"table-wrapper\\">
  <div className=\\"table-buttons\\">
    <Search text=\\"Search items by name\\" onChange={[Function: res]} />
    <SelectDropdown onChange={[Function: res]} name=\\"Name\\" userName=\\"Username\\" email=\\"Email\\" />
    <Btn onClick={[Function: onClick]} text=\\"Sort by ASC\\" />
  </div>
  <div className=\\"table-info\\">
    <div className=\\"users-wrapper\\">
      <h2>
        Users
      </h2>
      <div className=\\"user-keys-wrapper\\">
        <div className=\\"user-key-value\\">
          name
        </div>
        <div className=\\"user-key-value\\">
          username
        </div>
        <div className=\\"user-key-value\\">
          email
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Chelsey Dietrich
        </div>
        <div className=\\"user-info\\">
          Kamren
        </div>
        <div className=\\"user-info\\">
          Lucio_Hettinger@annie.ca
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Clementina DuBuque
        </div>
        <div className=\\"user-info\\">
          Moriah.Stanton
        </div>
        <div className=\\"user-info\\">
          Rey.Padberg@karina.biz
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Clementine Bauch
        </div>
        <div className=\\"user-info\\">
          Samantha
        </div>
        <div className=\\"user-info\\">
          Nathan@yesenia.net
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Ervin Howell
        </div>
        <div className=\\"user-info\\">
          Antonette
        </div>
        <div className=\\"user-info\\">
          Shanna@melissa.tv
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Glenna Reichert
        </div>
        <div className=\\"user-info\\">
          Delphine
        </div>
        <div className=\\"user-info\\">
          Chaim_McDermott@dana.io
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Kurtis Weissnat
        </div>
        <div className=\\"user-info\\">
          Elwyn.Skiles
        </div>
        <div className=\\"user-info\\">
          Telly.Hoeger@billy.biz
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Leanne Graham
        </div>
        <div className=\\"user-info\\">
          Bret
        </div>
        <div className=\\"user-info\\">
          Sincere@april.biz
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Mrs. Dennis Schulist
        </div>
        <div className=\\"user-info\\">
          Leopoldo_Corkery
        </div>
        <div className=\\"user-info\\">
          Karley_Dach@jasper.info
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Nicholas Runolfsdottir V
        </div>
        <div className=\\"user-info\\">
          Maxime_Nienow
        </div>
        <div className=\\"user-info\\">
          Sherwood@rosamond.me
        </div>
      </div>
      <div className=\\"user\\">
        <div className=\\"user-info\\">
          Patricia Lebsack
        </div>
        <div className=\\"user-info\\">
          Karianne
        </div>
        <div className=\\"user-info\\">
          Julianne.OConner@kory.org
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Table Table renders data correclty with provided props for Users and sort ASC 1`] = `
"<div className=\\"table-wrapper\\">
  <div className=\\"table-buttons\\">
    <Search text=\\"Search items by name\\" onChange={[Function: res]} />
    <SelectDropdown onChange={[Function: res]} name=\\"Name\\" userName=\\"Username\\" email=\\"Email\\" />
    <Btn onClick={[Function: onClick]} text=\\"Sort by DESC\\" />
  </div>
  <div className=\\"table-info\\">
    <div className=\\"users-wrapper\\">
      <h2>
        Users
      </h2>
      <div className=\\"user-keys-wrapper\\" />
    </div>
  </div>
</div>"
`;

exports[`Table Table renders data correclty with provided props for Users and sort DESC 1`] = `
"<div className=\\"table-wrapper\\">
  <div className=\\"table-buttons\\">
    <Search text=\\"Search items by name\\" onChange={[Function: res]} />
    <SelectDropdown onChange={[Function: res]} name=\\"Name\\" userName=\\"Username\\" email=\\"Email\\" />
    <Btn onClick={[Function: onClick]} text=\\"Sort by ASC\\" />
  </div>
  <div className=\\"table-info\\">
    <div className=\\"users-wrapper\\">
      <h2>
        Users
      </h2>
      <div className=\\"user-keys-wrapper\\" />
    </div>
  </div>
</div>"
`;
